crates/sui-core/src/authority/authority_store.rs

```diff
@@ -374,13 +375,27 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
                     }
                 }
                 InputObjectKind::ImmOrOwnedMoveObject(objref) => {
-                    if !self.object_exists(&objref.0, objref.1)? {
+                    if let Some(obj) = self.get_object_by_key(&objref.0, objref.1)? {
+                        if !obj.is_immutable() {
+                            probe_lock_exists.push(*objref);
+                        }
+                    } else {
                         missing.push(ObjectKey::from(objref));
                     }
                 }
             };
         }
 
+        if !probe_lock_exists.is_empty() {
+            match self.lock_service.locks_exist(probe_lock_exists).await {
```

Comment:
Because if the lock does not yet exist, the transaction can't be executed yet, so effectively it is the same as if the object didn't exist.

> Can the validation in lock service be done on the missing objects after get_missing_input_objects()?

I don't think I understand this question

Reply:
Basically I want to keep the semantics of get_missing_input_objects() simple. Intuitively missing locks are different from missing objects. If sui should guarantee that lock exists if object exists, treating lock not exists as a missing input object can complicate other parts of the system. For example, missing input objects can only be notified when the object commits. Is there the same guarantee for locks?

I know we have talked about the order of writing locks vs objects after a transaction is finished. If you think missing locks are expected behavior, and treating them as missing input objects are correct (they will be eventually notified), can you leave a detailed comment in the code?

The 2nd question is that to keep the logic simple, it seems better to call `get_missing_input_objects()` first, then scan objects that are not missing for locks.

Reply:
I verified that objects exist notifications are sent after writing locks. So I think what we need here is a comment why locks may not exist, and after locks are written TransactionManager will get notified.

Reply:
got it - left a comment!